### ISSUE TITLE
Confusion between 1-based and 0-based stream positions in different DB implementations of Event/Checkpoint Store

### ISSUE THREAD

**Issue by ReinderReinders** [Eventuous/eventuous#163]
I am working on a Proof-Of-Concept using Eventuous with PostgresDB as the Event Store, and Microsoft SQL Server as the Checkpoint Store / read model store.
I noticed the following confusing issue:

the global_position in the Postgres Schema is maintained as a 1-leading array:

![image](https://user-images.githubusercontent.com/119332127/204304653-251a4ab6-d4c1-4762-af78-dcd2e1834d79.png)

However the Subscription in the receiving SQL Server store tracks the _same_ global stream as a 0-leading array:

![image](https://user-images.githubusercontent.com/119332127/204304955-1a325fbf-c5a6-465e-bbd9-74d2b8a21f51.png)

I have double-checked and the Subscription tracks in real-time, is in sync, no event was lost or skipped etc., my 'receiving' application is not stalling. Triggered one more event after above screenshots to demonstrate this:

[Postgres Event Store]
![image](https://user-images.githubusercontent.com/119332127/204305591-4fb47ea7-0204-4863-9879-103dcdbdb596.png)
[SQL Server Checkpoint Store]
![image](https://user-images.githubusercontent.com/119332127/204305703-991015bf-62ac-47b0-bb5a-5a4bcb6d6f8b.png)

I find this confusing and not very intuitive. Am I missing some part of the implementation / intention here, or is this really just a little bug?

Kind regards,
Reinder

**Comment by alexeyzimarev** [comment]
I am not sure why that is since the checkpoint should be the global position value of the last processed event.

I would not expect the checkpoint store to manipulate the value in any way. It might be that the stored procedure that reads events from Postgres is doing something. Can you guess where the issue is coming from?

**Comment by ReinderReinders** [comment]
I'd guess maybe in the Eventuous.SqlServer.Subscriptions namespace, SqlServerAllStreamSubscription, AsContext method:

![image](https://user-images.githubusercontent.com/119332127/204312419-7532b9be-abe6-4979-8da0-937abaaa0048.png)

Looks like a hardcoded conversion from 1-leading to 0-leading to me?

edit: this implementation is identical to the Postgres Subscription implementation. I can't remember seeing this issue there before.. I'll run a test.

edit 2: The behaviour is identical (i.e. offset by one) using Postgres as the Checkpoint store:

![image](https://user-images.githubusercontent.com/119332127/204313643-d7ad5f67-da57-44ea-a8f2-63380b563ac9.png)

**Comment by alexeyzimarev** [comment]
Yeah, I can't remember now. I was cleaning up the Postgres version, and it was taken as a baseline for the MS SQL version. Indeed, we'd better confirm what's where. My guess is the global position number is produced by the sequence (Postgres) or identity enumerator in SQL Server, and it always starts with one, not zero. As we want to keep the first event as zero position, it requires such manipulation. So, inside the system, it is consistent (zero-leading).

**Comment by ReinderReinders** [comment]
Yes, you were right on the money. Postgres does not accept a 0-based sequence:

![image](https://user-images.githubusercontent.com/119332127/204528207-cb891625-ef54-4824-a017-b5f1c80599e2.png)

(I had hoped this would work but it is a technical impossibility: )

![image](https://user-images.githubusercontent.com/119332127/204528373-446745ca-3052-4623-8dfb-f83908f737a0.png)

Edit: Makes me wonder though, why should we want the first event as the zero position? This causes another issue that I will post shortly.
Edit2: This is the issue I am referring to: https://github.com/Eventuous/eventuous/issues/164

**Comment by alexeyzimarev** [comment]
That's what I meant. So, basically, the `+ 1` in the context construction just brings the sequencing to zero-lead as it's the default behaviour for Eventuous. It's confusing when you look at the database. For the rest - not so much. Maybe just mentioning it in the docs would suffice?

**Comment by alexeyzimarev** [comment]
With #164 presumably fixed, is this one still relevant?

**Comment by ReinderReinders** [comment]
@alexeyzimarev this issue is not specifically a bug, more like a feature :) I anticipate a hard time explaining to my colleagues why there is an offset between the global_position and the checkpoint position. But it is not breaking in any way so I guess we can live with it. It's just that I expect it to be confusing for (beginning) users of Eventuous. Adding a note to the documentation is of course always a good idea.
I understand the need to keep existing functionality intact (since Eventuous is already being used in production by others as far as I know), so this issue should not be corrected, but I was thinking we might make it configurable (making checkpoints 0 or 1 based I mean)? With the default value being the way it is now. That way users can choose their own preferred implementation.
But otherwise I would be all right with closing this issue.

**Comment by alexeyzimarev** [comment]
Yeah, it's a bit too late as changing the sequence would produce hidden consequences of skipping one event. I haven't been able to document both SQL stores yet, will keep in mind this issue.
